Updated bcounter to allow operations that have no id() (alias DCID) a…

FairPlayer4 opened the pull request:
…nd added proper test module

Operations that have no id() will be rejected from the downstream function with {error, no_permissions}
This is mainly intended to allow bcounter operations to pass the initial antidote operation check and another module (bcounter_mgr) makes sure that operations get an id() added
Additionally more tests were added and preliminary proper tests for counter_b were added
Also added some more documentation

This builds upon what has been proposed in PR #29 by @SmallEndian and extends it to all updates in the antidote_crdt_counter_b

FairPlayer4 committed:
Updated bcounter to allow operations that have no id() (alias DCID) and added proper test module
Operations that have no id() will be rejected from the downstream function with {error, no_permissions}
This is mainly intended to allow bcounter operations to pass the initial antidote operation check and another module (bcounter_mgr) makes sure that operations get an id() added
Additionally more tests were added and preliminary proper tests for counter_b were added

bieniusa commented on src/antidote_crdt_counter_b.erl:
```diff
@@ -220,6 +237,8 @@ equal(BCounter1, BCounter2) ->
 
 -ifdef(TEST).
```
Please keep all includes at the top of the file. This is consistently done in all other files.

FairPlayer4 replied:
Done

bieniusa commented on test/prop_counter_b.erl:
```diff
@@ -0,0 +1,102 @@
+            not antidote_crdt_counter_b:is_operation(Operation)
+        end).
+
+prop_increment_decrement_test() ->
```
Comments would be helpful here to understand the general structure of the test.

FairPlayer4 replied:
Added some comments

bieniusa reviewed (changes requested):
Looks good - thank you! Just a few tiny things.

peterzeller commented:
Opinion: I guess you are making the `antidote_crdt_counter_b_op` less precise to avoid some of the special cases for bounded counters in the Antidote code base? I would prefer to keep the type precise here and always require the actor id. The CRDT library should also make sense as a standalone library and as such an increment without an actor is not a valid operation for the bounded counter.

Regarding the property tests: 
I think you are not winning anything by using random values for these tests. For `prop_is_operation_test` and `prop_is_not_operation_test` you can easily use exhaustive testing and test all the possible cases via eunit tests if you just test -1, 0, and 1 for the numbers.
Similarly for `prop_is_not_operation_test` you just have a few interesting equivalence classes so in my opinion it would be clearer to use fixed inputs here.
Random executions could be interesting if you had more than two operations with several actor-ids and transfers.

FairPlayer4 committed:
Made requested changes
Added more extensive proper test

FairPlayer4 committed:
Added another eunit test

FairPlayer4 committed:
Fixed lint issue hopefully

FairPlayer4 committed:
Added more documentation and updated existing docs
Added alternative_value function for review
Added small proper test

FairPlayer4 commented:
@peterzeller 
Thanks for the feedback,
This is just one alternative, I could make the change in the Antidote code base. 
But I think this change has less of a performance impact (less specific checking in Antidote) and is easier to understand for clients and developers.
An undocumented and non-existing operation in the antidote_crdt library will not be used by the client and will only pollute the Antidote codebase and confuse the developers because they don't understand why the special case exists in the first place.
If the operation exists in the antidote_crdt library it is clear that it is allowed and can be used.
Other developers that want to use our antidote_crdt library will have to handle the {error, no_permissions} case anyway so the extra operations in antidote_crdt_counter_b do not impact them. They can choose whether to support them or not.
I added more documentation so that other developers can see how they have to handle the new operations.
I would also argue that it is much easier to test this change in the antidote_crdt library because the code is not connected to other modules.
Lastly, I think the library is intended mainly for Antidote so making a few changes to fit our use case more is not a bad idea.

Regarding proper:
I just got started with proper testing and just played around with simple tests
I added operation combinations (positive and negative) as part of the eunit tests so I could probably remove those proper tests if that is required (I assume they don't really hurt here)
I tried to test everything with eunit tests and then just add proper tests on top of that
I added some more proper tests of which the last one tests every operation and applies it

> 
> 
> Opinion: I guess you are making the `antidote_crdt_counter_b_op` less precise to avoid some of the special cases for bounded counters in the Antidote code base? I would prefer to keep the type precise here and always require the actor id. The CRDT library should also make sense as a standalone library and as such an increment without an actor is not a valid operation for the bounded counter.
> 
> Regarding the property tests:
> I think you are not winning anything by using random values for these tests. For `prop_is_operation_test` and `prop_is_not_operation_test` you can easily use exhaustive testing and test all the possible cases via eunit tests if you just test -1, 0, and 1 for the numbers.
> Similarly for `prop_is_not_operation_test` you just have a few interesting equivalence classes so in my opinion it would be clearer to use fixed inputs here.
> Random executions could be interesting if you had more than two operations with several actor-ids and transfers.

albsch commented:
Please reopen against https://github.com/AntidoteDB/antidote
